### ISSUE TITLE
Feature request: mark parenthesized expressions in AST

### ISSUE THREAD

**Issue by thorn0** [typescript-eslint/typescript-eslint#1682]
**What code were you trying to parse?**

```ts
/** @type {Date} */ (expr)
```

**What did you expect to happen?**

There are cases when AST consumers need to know whether an expression had parens. Babel adds `.extra.parenthesized: true` to nodes. My proposal is to do the same thing. For example, this feature would be useful for Prettier to format [Closure-style type casts](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#type-casting) properly.

**What actually happened?**

The parser ignores parens when building the AST.

**Versions**

not relevant

**Comment by bradzacher** [comment]
I like the idea behind this, but my problem is that it's a deviation from the ESTree spec.

No doubt the entire ecosystem would be able to benefit from this (saves eslint rules having to do token checks).

So maybe it makes sense to write up a proposal for the spec?

**Comment by thorn0** [comment]
Related ESTree issue: https://github.com/estree/estree/issues/194

Also I later realized that for the needs of Prettier `.extra.parenthesized: true` doesn't suffice because it can't represent nested parens. That's why Prettier with `--parser babel` uses `ParenthesizedExpression` now.

**Comment by armano2** [comment]
i'm going to add this option today

**Comment by bradzacher** [comment]
This should be pretty simple to implement.
For example:

```ts
if (node.kind === ts.SyntaxKind.ParenthesizedExpression) {
  if (ts.getJSDocTypeTag(node) != null) {
    estreeNode = <JSDocTypeAssertion>;
  }
}
```

```ts
interface JSDocTypeAssertion {
  type: AST_NODE_TYPES.JSDocTypeAssertion;
  expression: Expression;
}
```

I don't think we need to include the `@type` in the comment (we don't yet support JSDoc comment parsing anywhere, so I don't think it's a good idea to do that yet).

**Comment by thorn0** [comment]
@armano2 Are you going to add the `parenthesized` property to existing nodes? As I wrote this isn't really useful for Prettier. I suppose it won't be useful for ESLint rules either. `ParenthesizedExpression` is a better thing. I guess it's fine to create these nodes only if a `@type` comment is present.

**Comment by armano2** [comment]
> Are you going to add the parenthesized property to existing nodes

no, this is bad idea

> ParenthesizedExpression is a better thing

`ParenthesizedExpression` node is going to be created, but its going to be hidden behind compiler flag, that we don't have to make breaking change for this

if ParenthesizedExpression has comments or direct child is a comment we can create it, typescript already creates this node and we are just skipping it.

https://github.com/typescript-eslint/typescript-eslint/blob/3e204e9f2fe7008c7d7127d865f58bff1d978e0a/packages/typescript-estree/src/convert.ts#L2317-L2318

**Comment by bradzacher** [comment]
> that we don't have to make breaking change for this

We won't ever get turned on by default - that's a huge breaking change from ESTree spec and would break most lint rules.

**Comment by bradzacher** [comment]
It was closed because it had been open for an entire year without a response from the author to the review

We are a community-run project. The volunteer maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues will not progress unless a member of the community steps up and champions it.

If this issue is important to you — consider being that champion!

**Comment by jwbth** [comment]
Now that JSDoc allows to do _almost everything_ TypeScript does including generics (see [the reference](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) for TypeScript-flavored JSDoc), and [recently even function overloads](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#@overload-support-in-jsdoc), there is a huge niche for coding in pure JS with type safety. Moreover, TypeScript extensions in IDEs like VS Code support this. So, this is a gap in typescript-eslint compared to IDE-based type checking.

I develop user scripts for Wikimedia projects (and MediaWiki in general), and due to our infrastructure we write scripts in JS. We can't write in TS, except with using a build step!
1. The majority of user script maintainers can't afford that. 
2. That negates all the advantages that the MediaWiki script infrastructure provides (e.g. [ResourceLoader](https://www.mediawiki.org/wiki/ResourceLoader)).

I believe we deserve type safety with state-of-the-art solutions like `typescript-eslint` as well. The absence of support for the requested feature complicates type casting operations. Basically, you have to always introduce a new variable with a type-casted value. It's bad enough that we have to write `/** @type {Type} */ (obj)` instead of `(obj as Type)`. But we also should introduce a variable each time something has to be type-casted:
```js
const typeCastedObj = /** @type {Type} */ (obj);
typeCastedObj  // operations with typeCastedObj
```

I would happily write the PR myself, but unfortunately the internals of `typescript-eslint` and TypeScript's AST are a completely unfamiliar area to me. So I urge anyone with the necessary expertise to help introduce support for the feature.

**Comment by bradzacher** [comment]
> there is a huge niche

That there is the caveat -- it is a niche.
The vast, vast, vast majority of our users write `.ts` files.

There's a large cost to supporting this in the parser.
There's a large cost to supporting this in lint rules.

We are a volunteer run project and the maintainers are barely able to allocate enough of their free time to this project to meet the current maintenance needs.

The reality is that there simply is not enough value (as in, enough users would benefit from the support) to justify the cost of adding 1st-class support for JSDoc assertions.

----

This issue is specifically about adding parser support so that projects like prettier could opt-in to the AST and thus make it easier for them to do their formatting. This does not include adding support into lint rules.
i.e. this is not a feature request that impacts our ESLint users -- it is a feature request that impacts our parser users.

----

Perhaps it would be a good idea to talk to MediaWiki about adding support for typescript to their ResourceLoader. Looking at the page you linked it mentions it already does some transforms on the code provided -- so it seems logical that they could also do type stripping (eg the same thing NodeJS now does for `.ts` files) to enable use of TS in scripts.

**Comment by jwbth** [comment]
@bradzacher Thanks for the reply.

> This issue is specifically about adding parser support so that projects like prettier could opt-in to the AST and thus make it easier for them to do their formatting. This does not include adding support into lint rules. i.e. this is not a feature request that impacts our ESLint users -- it is a feature request that impacts our parser users.

But is it a first step to implementing said support?

> Perhaps it would be a good idea to talk to MediaWiki about adding support for typescript to their ResourceLoader. Looking at the page you linked it mentions it already does some transforms on the code provided -- so it seems logical that they could also do type stripping (eg the same thing NodeJS now does for `.ts` files) to enable use of TS in scripts.

I agree with you that such support would be nice. But it's not even planned yet. Based on what I know, it is _much_ easier said than done, and I see many complications that would come up.

MediaWiki is, to a large degree, a volunteer-run project as well (I'm personally a volunteer developing user scripts, and this is volunteers who are to benefit from this feature), and ultimately it's coordinated by a non-profit funded by donations. So, our limitations are comparable at that...
